### PR TITLE
Maximize usage of available space in tile view

### LINKED ISSUE #5142: [UI] Tile view does not act nice with narrow and tall windows
Steps to reproduce:
* Join a jitsi call with 6 people
* Enable tile view
* Make the screen narrow and tall

Expected behaviour:
* The tile view fills the available space in a 1-column 6-row grid.

Actual behaviour
* The tile view only uses a small portion of the available screen space in a 3-column 2-row grid.

### PR DESCRIPTION
**Pull request by setpill**
Fixes #5142

### CODE REVIEW

**Comment by setpill** [comment]
I do not want to sign any documents with my full name and address and doxx myself just to contribute to an open source project. I consent to my contribution becoming part of the `apache 2.0`-licensed codebase, with all of the implications that come from that.

**Comment by dfrumin** [comment]
Hi Jitsi team - is there any way to unblock this fix?  Maybe someone else could be inspired by setpill's code and write it / sign the docs?  This is precisely the issue we've been having the last couple of weeks so I'm excited to see a fix get merged in...  THANKS!!!!

**Comment by saghul** [comment]
I’ll try to take a look in the next few days. Sorry for the delay!

File: react/features/video-layout/functions.js [hunk 1]
```diff
@@ -35,30 +35,49 @@ export function getMaxColumnCount() {
...
  * rows, and visible rows (the rest should overflow) for the tile view layout.
  */
-export function getTileViewGridDimensions(state: Object, maxColumns: number = getMaxColumnCount()) {
+export function getTileViewGridDimensions(state: Object) {
     // When in tile view mode, we must discount ourselves (the local participant) because our
     // tile is not visible.
     const { iAmRecorder } = state['features/base/config'];
     const numberOfParticipants = state['features/base/participants'].length - (iAmRecorder ? 1 : 0);
+    const { clientHeight, clientWidth } = state['features/base/responsive-ui'];
+    const aspectRatio = state['features/base/responsive-ui'].aspectRatio;
```
Comment (saghul) [inline react/features/video-layout/functions.js:51]: you can extract this too in the above line

File: react/features/video-layout/functions.js [hunk 2]
```diff
@@ -35,30 +35,49 @@ export function getMaxColumnCount() {
 }
 
 /**
- * Returns the cell count dimensions for tile view. Tile view tries to uphold
- * equal count of tiles for height and width, until maxColumn is reached in
- * which rows will be added but no more columns.
+ * Returns the cell count dimensions for tile view. Tile view tries to fill
+ * the available space optimally
```
Comment (saghul) [inline react/features/video-layout/functions.js:39]: end the sentencce in a period or the linter complains

File: react/features/video-layout/functions.js [hunk 3]
```diff
@@ -35,30 +35,49 @@ export function getMaxColumnCount() {
...
+    let bestFound = false;
+
+    while (!bestFound) {
+        const columns = bestColumns + 1;
+        const straightTileWidth = Math.floor(clientWidth / columns);
+
+        const rows = Math.ceil(numberOfParticipants / columns);
+        const straightTileHeight = Math.floor(clientHeight / rows);
+
+        const constrainedTileWidth = Math.ceil(straightTileHeight * aspectRatio);
```
Comment (saghul) [inline react/features/video-layout/functions.js:65]: aspectRatio is a Symbol. Have you tested this code?

File: react/features/video-layout/functions.js [hunk 4]
```diff
@@ -35,30 +35,49 @@ export function getMaxColumnCount() {
...
+            bestTileWidth = tileWidth;
+        } else {
+            bestFound = true;
+        }
+    }
 
     return {
-        columns,
-        visibleRows
+        bestColumns,
```
Comment (saghul) [inline react/features/video-layout/functions.js:79]: Changing the keys of the returned object surely breaks the code using it.

**Comment by saghul** [comment]
Made a review round and left some comments!

**Comment by setpill** [comment]
Thanks for the review, will take a look at the comments.

File: react/features/video-layout/functions.js (hunk 3 above)
Comment (setpill) [reply react/features/video-layout/functions.js:65]: No, I have not been able to do so, as I do not have a dev environment set up for this codebase. It was much quicker to just whip up some code than to get the dev env set up. When I was faced with the bot telling me to sign a legal document it also demotivated me from putting in that work as I wasn't sure you would even consider this PR. Your review gives me indication that it's worth it after all to look into that, so hopefully I'll get it working soon :)

File: react/features/video-layout/functions.js (hunk 4 above)
Comment (setpill) [reply react/features/video-layout/functions.js:79]: Aren't these just the variable names that are inserted into the returned array? I am not that familiar with js, but this doesn't look like a dictionary or mapping of some kind.

**Comment by dfrumin** [comment]
Thanks you guys.  I'm bummed that I'm a bit tied up or else it would be fine to set up the code and play in it.  Maybe someday soon.  :-)

File: react/features/video-layout/functions.js (hunk 4 above)
Comment (saghul) [reply react/features/video-layout/functions.js:79]: That's indeed a dictionary / object literal. Notice the curly braces.

File: react/features/video-layout/functions.js (hunk 3 above)
Comment (saghul) [reply react/features/video-layout/functions.js:65]: Thank you!

**Comment by JInnace** [comment]
Hey, I think I have a solution for this I'm formatting it to pass the linter now.

File: react/features/video-layout/functions.js (hunk 4 above)
Comment (setpill) [reply react/features/video-layout/functions.js:79]: Ah. I was not aware of this shorthand property name thing. Annoying language construct, obscuring the common syntactical signature of a dictionary, so I didn't recognize it as such. I will fix.

File: react/features/video-layout/functions.js (hunk 1 above)
Comment (setpill) [reply react/features/video-layout/functions.js:51]: Coding javascript feels like throwing spaghetti at the wall and seeing what sticks. Guesstimated a thing that somehow might magically work. No clue.

Comment (setpill) [reply react/features/video-layout/functions.js:51]: I used the wrong aspectRatio. Should be fixed now.

File: react/features/video-layout/functions.js (hunk 3 above)
Comment (setpill) [reply react/features/video-layout/functions.js:65]: Resolving this thread as I have changed the aspectRatio referenced. Will still look into dev/testing env.

**Comment by setpill** [comment]
Just an update: I have not yet been able to fix a dev env for myself and I am not 100% sure when I will have time to look into that.

**Comment by MurzNN** [comment]
Hey, what is the status of this PR? Tiles in low height are not usable in Jitsi

**Comment by setpill** [comment]
I don't think I'll be able to work on this any time soon. Anyone who wants to use my contributions to base their own PR on, please feel free to do so.

**Pull request closed without merging**